service/history/replicationTaskProcessor.go

```diff
@@ -0,0 +1,285 @@
+				token:    &r.ReplicationToken{ShardID: common.Int32Ptr(int32(p.shard.GetShardID())), TaskID: common.Int64Ptr(p.readLevel)},
+				respChan: respChan,
+			}
+
+			select {
+			case response := <-respChan:
+				p.logger.Info("Got fetch replication tasks response.",
+					tag.ReadLevel(response.GetReadLevel()),
+					tag.Bool(response.GetHasMore()),
+					tag.Counter(len(response.GetReplicationTasks())),
+				)
+
+				for _, replicationTask := range response.ReplicationTasks {
+					p.processTask(replicationTask)
+				}
+
+				p.readLevel = response.GetReadLevel()
+				err := p.shard.UpdateClusterReplicationLevel(p.sourceCluster, p.readLevel)
+				if err != nil {
+					p.logger.Error("Error updating replication level for shard", tag.Error(err), tag.OperationFailed)
+				}
+
+				scope.UpdateGauge(metrics.ReplicationTasksReadLevel, float64(p.readLevel))
+				scope.AddCounter(metrics.ReplicationTasksApplied, int64(len(response.GetReplicationTasks())))
+			case <-p.done:
+				p.logger.Info("Closing replication task processor.", tag.ShardID(p.shard.GetShardID()), tag.ReadLevel(p.readLevel))
+				return
+			}
+		}
+	}()
+
+	p.logger.Info("ReplicationTaskProcessor started.")
+}
+
+func (p *replicationTaskProcessor) Stop() {
+	close(p.done)
+}
+
+func (p *replicationTaskProcessor) processTask(replicationTask *r.ReplicationTask) {
+	var err error
+SubmitLoop:
+	for {
+		var scope int
+		switch replicationTask.GetTaskType() {
+		case r.ReplicationTaskTypeDomain:
+			scope = metrics.DomainReplicationTaskScope
+			err = p.handleDomainReplicationTask(replicationTask)
+		case r.ReplicationTaskTypeSyncShardStatus:
+			scope = metrics.SyncShardTaskScope
+			err = p.handleSyncShardTask(replicationTask)
+		case r.ReplicationTaskTypeSyncActivity:
+			scope = metrics.SyncActivityTaskScope
+			err = p.handleActivityTask(replicationTask)
+		case r.ReplicationTaskTypeHistory:
+			scope = metrics.HistoryReplicationTaskScope
+			err = p.handleHistoryReplicationTask(replicationTask)
+		case r.ReplicationTaskTypeHistoryMetadata:
+			// Without kafka we should not have size limits so we don't necessary need this in the new replication scheme.
+		default:
+			p.logger.Error("Unknown task type.")
+			scope = metrics.ReplicatorScope
+			err = ErrUnknownReplicationTask
+		}
+
+		if err != nil {
+			p.logger.Info("Failed to apply replication task.", tag.TaskID(replicationTask.GetSourceTaskId()))
+			p.updateFailureMetric(scope, err)
+			if !isTransientRetryableError(err) {
+				p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster)).IncCounter(metrics.ReplicationTasksFailed)
+				break SubmitLoop
+			}
+		} else {
+			p.logger.Info("Successfully applied replication task.", tag.TaskID(replicationTask.GetSourceTaskId()))
+			p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster)).IncCounter(metrics.ReplicationTasksApplied)
+			break SubmitLoop
+		}
+	}
+
+	if err != nil {
```

Comment:
I think you may want to return an error to the caller.

Reply:
Caller cannot do anything in this case. We can at best put in our own DLQ and then it will require human intervention.